Technical Specifications and Software Status Final Report Updates (#522)
jpabadir/pace-c: documentation/final-deliverables/pdf-report.md

## Technical Specifications

**Overview:**

> Although the decision for our chosen tech stack was ultimately up to us, our client has technical experience, and so she offered recommendations for front-end and back-end libraries back in September. As such, we aimed to align with the client’s technical suggestions, as well as to pick an MVC model which was well documented. Ultimately, this resulted in React being used for our front-end, Node for our back-end, and Firebase for our database.

Reviewer: I don't think React uses the MVC model. From [their doc](https://reactjs.org/blog/2013/06/05/why-react.html): "React isn’t an MVC framework."
```suggestion
Although the decision for our chosen tech stack was ultimately up to us, our client has technical experience, and so she offered recommendations for front-end and back-end libraries back in September. As such, we aimed to align with the client’s technical suggestions, as well as to pick well-documented technologies. Ultimately, this resulted in React being used for our front-end, Node for our back-end, and Firebase for our database.
```

Unless you mean that the combo of these three things is MVC?

Author: Yea, my thinking was that the combination of React, Node, and Firebase would follow an MVC model, but my logic could be flawed! I'll commit your suggestion :)

**Technical Specifications on the Front-end:**

- React is a library that utilizes JavaScript to create user interfaces.
- Code pertaining to the front-end can be accessed by navigating to the ‘src’ folder, then by clicking on the ‘front-end’ folder.
- ‘App.js’ is the main page that launches upon the application being run.

>   - Every other UI page is split into components that are rendered dependent on the program’s state from the user’s actions. Given this, all of our UI components exist in their own folder named according to what it displays within the central folder titled ‘components’.

Reviewer: ```suggestion
  - Every other UI page is split into components that are rendered dependent on the program’s state from the user’s actions. Given this, all of our UI components exist in their own folder named according to what they display within the central folder titled ‘components’.
```

- ‘helper-methods.js’ contains various functions that are called throughout the application. The majority of these functions play a crucial role for interacting with the Firebase database (for example, to create users after form submission, to send verification and password reset emails for authenticated users, etc.).
- Credentials for interacting with the database are within firebase-init.js.

> - ‘index.js’ allows the Mentee and Mentor forms to be rendered independently from the central website.

Reviewer: ```suggestion
- ‘index.js’ allows the Mentee and Mentor forms to be rendered independently from the rest of the website. This is useful when we need to show only the mentor form to a mentor we just invited via email, and when we need to embed only the mentee form in another organization's website. 
```

- ‘timeZones.json’ is utilized to populate the ‘time zone’ dropdown on the Mentee and Mentor forms.

A major benefit to using React is the ability to use UI Libraries (we opted for Ant Design), as well as to integrate with our back-end Node server. The main Ant Design components utilized throughout the app are Notifications, Cards, Forms, Icons, Tabs, and Modals.

**Technical Specifications on the Back-end:**

- Node is a back-end environment which uses JavaScript.
- Our back-end is accessed from the ‘back-end’ folder within ‘src’.

> - ‘App.js’ contains functions that are triggered from events on a specific port. For example, functions exist for generating matches between Mentees and Mentors (by finding the least difference in time zones, number of commons skills, and same organization string) as well as route calls for inviting new mentors and accepting mentees.

Reviewer: ```suggestion
- ‘app.js’ contains functions that are triggered from requests on specific ports. For example, functions exist for generating matches between Mentees and Mentors (by finding the least difference in time zones, number of commons skills, and same organization string) as well as route calls for inviting new mentors and accepting mentees.
```

- ‘Email-helpers.js’ houses functions that send custom emails to users via nodemailer.

> - The back-end is configured to use the Firebase Admin SDK as a way of securing user information.

Reviewer: ```suggestion
- The back-end is configured to use the Firebase Admin SDK. This is needed to give our backend full access to our firebase database, which is, otherwise, secured by rules. 
```

Note that both the front-end and back-end contain files to assist in catching errors (.eslintrc.json) and for automatic formatting adjustments (.prettierrc.json). The package-lock.json files contain the list of dependencies which are installed whenever the ‘npm install’ command is run in the terminal before running the front-end and back-end.

**Technical Specifications of the Database:**

- Firebase is a real-time NoSQL database owned by Google, which is hosted on the cloud.
- By navigating to the ‘Realtime Database’ tab of the Firebase console, the JSON data can be referenced. Two nodes exist in our database’s structure:

1.  Organizations - Allow multiple Admins to exist within an Organization, as well as contain mentor email addresses who are either active or pending (for clarity: activeMentors and pendingMentors are ‘children’ of the ‘Organization’ node).
2.  Users – Contains unique user IDs for each type of user (Admin, Mentor, or Mentee), with the information (i.e ‘children’ of the ‘Users’ node) gathered from the form used to register.

> - The ‘Authentication’ tab of the Firebase console allows us to verify Mentor and Admin identities upon account creation as well as to validate credentials upon login. All users are required to verify their email address before being able to access Admin or Mentor privileges of the website.

Reviewer: ```suggestion
- The ‘Authentication’ tab of the Firebase console allows us to verify Mentor and Admin identities upon account creation as well as to validate credentials upon login. All users are required to verify their email address before being able to access Admin or Mentor privileges of the website. Note that the user id of each user in the 'Authentication' menu is the same as the user id of each user found under the 'Realtime Database' menu.
```

  - Our Firebase configuration uses the ‘Email/Password’ Sign-in method. This can be modified if needed in the future by going to the ‘Sign-in method’ tab under ‘Authentication’.

>   - Authentication also handles password reset changes.

Reviewer: ```suggestion
  - Authentication also handles password reset changes.
  - Since we use Firebase's 'Authentication' service, we never have to worry about encrypting or storing users' passwords, because that is also handled for us. 
```

Author: Ah, right! Thanks for adding this in. I remember asking about the password visibility back in October, so this is a really great point to incorporate

<br>

## Limitations of MentorMatch (and future improvements)

- The way that a mentor's availability is obtained is through a url input field on the Mentor form. This field is currently limited to only accept Google Calendar links.

- Mentees and Mentors communicate via email messages externally from MentorMatch. Due to this, if MentorMatch was further developed to require internal communication between Mentors and Mentees, Mentees would also need a MentorMatch account.

- Mentors will have to manually update their calendar to reflect booked timeslots with Google Calendars. _Note_: (Google Appointment Slots)[https://support.google.com/calendar/answer/190998?co=GENIE.Platform%3DDesktop&hl=en] would help to remediate this, but since we don't have access to an organization or school Google Calendar, it is not something that we are able to test. However, organizations will likely be able to use Appointment Slots.

> - In the case that MentorMatch grows and needs to send >500 emails per day, Nodemailer will become a limitation.

Author: @jpabadir I realized after I made these changes that you'll be working on the '## Limitations of MentorMatch (and future improvements)' of the document. I hope that the changes that I've added don't over-step too much on your task!

Reviewer: This is great, Katrina. Thanks for helping me get ahead :)
